libs/ktem/ktem/tags/crud.py

```diff
@@ -0,0 +1,244 @@
+# mypy: ignore-errors
+
+from collections import defaultdict
+from datetime import datetime
+
+from ktem.db.base_models import BaseTag, TagProcessStatus, TagType
+from ktem.db.models import ChunkTagIndex, Tag
+from sqlmodel import Session, select
+
+
+class TagCRUD:
+    def __init__(self, engine):
+        self._engine = engine
+
+    def list_all(self) -> list[dict]:
+        with Session(self._engine) as session:
+            statement = select(Tag)
+            results = session.exec(statement).all()
+
+        records: list[dict] = []
+        for result in results:
+            records += [dict(result)]
+
+        return records
+
+    def create(
+        self,
+        name: str,
+        prompt: str,
+        config: str,
+        type: str = "text",
+        scope: str = "chunk",
+        valid_classes: str | None = None,
+    ) -> str | None:
+        meta = {}
+        if type == TagType.classification:
+            assert (
+                valid_classes is not None and valid_classes != ""
+            ), "Invalid valid classes!"
+            meta["valid_classes"] = valid_classes
+
+        with Session(self._engine) as session:
+            # tag name must be unique
+            existed_result = self.query_by_name(name)
+            if existed_result:
+                raise Exception(f"Tag name: {name} has already been existed!")
+
+            tag: Tag = Tag(
+                name=name,
+                prompt=prompt,
+                config=config,
+                type=type,
+                scope=scope,
+                meta=meta,
+            )
+            session.add(tag)
+            session.commit()
+
+            return tag.id
+
+    def query_by_id(self, tag_id: str) -> BaseTag | None:
+        with Session(self._engine) as session:
+            statement = select(Tag).where(Tag.id == tag_id)
+
+            result = session.exec(statement).first()
+            return result
+
+    def query_by_ids(self, tag_ids: list[str]) -> list[dict]:
+        with Session(self._engine) as session:
+            statement = select(Tag).where(Tag.id.in_(tag_ids))
+
+            results = session.exec(statement).all()
+            results = [dict(result) for result in results]
+        return results

```

Comment:
this is more recommended to return a list of `Model` than just returning a list of `dict`. The model itself is more structurized